fixed `dotnet sln add` blobs

gforceg opened the pull request:
# fixed `dotnet sln add` blobs

## Summary

No projects are added to the solution file when the user runs `dotnet sln todo.sln add **/**` if there are any sub-directories that do not contain a project file. For example, the presence of  wwwroot/ in an API project will cause the command to fail. Most solutions contain projects that have sub-folders.

Using `dotnet sln todo.sln add **/**.{csproj,fsproj}` does work.

## Details

Under bash 4.3.42(1) `dotnet sln add` fails when the wildcard blob does not include a file extension qualifier (e.g., .csproj)

`dotnet sln <FILE.sln> add **/*`

The CLI terminates when it comes across a sub directory that does not contain a project file. E.g., wwwroot/.
Adding a qualifier to the blob, (.{csproj,fsproj}) seems to be the correct way to use blobs with `dotnet sln add`.

This does not apply to `dotnet sln remove` which can use unqualified wildcard blobs.

gforceg committed:
fixed blobs in readme

Under bash 4.3.42(1) `dotnet sln add` fails when the wildcard blob does not include a file extension qualifier (e.g., .csproj)

`dotnet sln <FILE.sln> add **/*`

The CLI terminates when it comes across a sub directory that does not contain a project file. E.g., wwwroot/.
Adding a qualifier to the blob, (`.{csproj,fsproj}) seems to be the correct way to use blobs with `dotnet sln add`.

This does not apply to `dotnet sln remove` which can use unqualified wildcard blobs.

dnfclas commented:
This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

GuardRex commented:
@piotrpMSFT Could you elaborate on ...

> globbing is no longer accepted as an input to CLI commands

... from the discussion in https://github.com/dotnet/cli/issues/3409 as it pertains here?

Help on the command makes no indication that globbing will work at all:

> <SLN_FILE>  Solution file to operate on. If not specified, the command will search the current directory for one.

It also seems strange to me that the globstar is repeated in a couple of them (`**/**`), since I thought a single globstar would match all files in all directories and sub-directories when not followed by a pattern matching segment or a trailing `/`. Isn't just `**` sufficient?

Beyond that, I have a few mechanical questions. Should they be ...
```
dotnet sln [<SOLUTION_NAME>] add **[/*.csproj|/*.fsproj]
```
and
```
add **[/*.csproj|/*.fsproj]
```
denoting that the filenames and extensions are optional and sticking with our ordinary command notation? The examples can show the globbing patterns, but I think the command syntax should stick with the normal notation. *This all assumes that these patterns are even valid in the first place.*

mairaw commented:
@livarcocc can you review this PR and check @GuardRex's questions?

livarcocc commented:
This is not a feature of the CLI. This is a feature of bash. The CLI is not the one resolving the globing in this case. This does not work on windows, for instance. So, I don't think we should document this as being a CLI feature.

GuardRex commented:
@gforceg Let's consider that it might be nice to harden the doc a bit against deltas in glob behavior across platforms. The body can show a placeholder for the pattern, and the doc already points the reader to info on patterns. Therefore, one way to go (just discussing it here) is to have the synopsis section look like this ...
```
dotnet sln [<SOLUTION_NAME>] add <PROJECT> <PROJECT> ...
dotnet sln [<SOLUTION_NAME>] add <GLOBBING_PATTERN>
dotnet sln [<SOLUTION_NAME>] remove <PROJECT> <PROJECT> ...
dotnet sln [<SOLUTION_NAME>] remove <GLOBBING_PATTERN>
dotnet sln [<SOLUTION_NAME>] list
dotnet sln [-h|--help]
```
and then in `add` commands under the `<PROJECT>` one ...
```
add <GLOBBING_PATTERN>
```
with this under the `<PROJECT>` one for `remove` ...
```
remove <GLOBBING_PATTERN>
```
The examples can show actual patterns. Isn't the one you show ...
```
dotnet sln todo.sln add **/*.{csproj,fsproj}
```
... the same as ...
```
dotnet sln todo.sln add **
```
... if that will work?

Perhaps one with a globstar in the path would be cool ...
```
dotnet sln todo.sln add /folder/**/*.csproj
```
... which should get only *.csproj* file-based projects in any subdirectory under *folder*. I'm just thowing that out for discussion btw ... there might be better example(s) to use.

Also, would you mind updating the `ms.date` to today's date on your next pass?

GuardRex commented:
@gforceg ... any thoughts on my :point_up: comments?

gforceg commented:
@GuardRex,

Yes, you're right. The place holders make better sense because we are documenting the CLI not the CLI as it works in my own environment. :penguin:  

And I like your synopsis. I think it makes it clear to the reader that the globbing pattern examples aren't to be taken literally.

I'll make some changes and push to the pr.

>Isn't the one you show ...
> `dotnet sln todo.sln add **/*.{csproj,fsproj}`
>... the same as ...
> `dotnet sln todo.sln add **`

The problem I experienced with simply passing `**` to `dotnet-sln-add` (see the issue summary) is that the command fails because there are some subdirectories that do not contain a project file. (e.g., 'wwwroot/') The command fails before any projects can be added to the solution.

Maybe this is a but w/ `dotnet-sln-add`?

gforceg commented:
I tried `**` and it made no difference. Also I deteremind that `**/*` and `**` are not equivalent.
Created an issue on dotnet/cli because I'm thinking it is an issue w/ the CLI given that `dotnet-sln-remove **/*` does not choke like `dotnet-sln-add **/*` does.

gforceg committed:
Update dotnet-sln.md

change ms.date to now
replace specific globs with place holders in synopsis
removed 'add a new solution to the current directory' this is covered by dotnet-new documentation
provide examples for:
  adding / removing individual solutions
  adding / removing multiple solutions
  adding / remove multiple solutions using globbing patterns

GuardRex commented on docs/core/tools/dotnet-sln.md:
```diff
@@ -65,18 +63,26 @@ Prints out a short help for the command.
+
+Add multiple C# projects to a solution using a globbing pattern:
+
+`dotnet sln todo.sln add **/.csproj`
```
Did you intend to put the wildcard in there?
```
dotnet sln todo.sln add **/*.csproj
```

gforceg replied:
yes

GuardRex commented on docs/core/tools/dotnet-sln.md:
```diff
@@ -65,18 +63,26 @@ Prints out a short help for the command.
+Remove multiple C# projects from a solution using a globbing pattern:
 
-`dotnet sln add **/**/*.fsproj`
+`dotnet sln todo.sln remove **/.csproj`
```
Same question here ...
```
dotnet sln todo.sln remove **/*.csproj
```

GuardRex commented on docs/core/tools/dotnet-sln.md:
```diff
@@ -37,15 +37,13 @@ The `dotnet sln` command provides a convenient way to add, remove, and list proj
+`remove **/*.csproj`
 
-Remove a project or multiple projects from the solution file. [Globbing patterns](https://en.wikipedia.org/wiki/Glob_(programming)) are supported on Unix/Linux based terminals.
+Removes a C# project or multiple C# projects from the solution file. [Globbing patterns](https://en.wikipedia.org/wiki/Glob_(programming)) are supported on Unix/Linux based terminals.
```
I suggest reverting this back.

gforceg replied:
Should we keep 'Removes' but drop 'C#'?
The command also removes F# projects too.

gforceg commented on docs/core/tools/dotnet-sln.md:
```diff
@@ -37,15 +37,13 @@ The `dotnet sln` command provides a convenient way to add, remove, and list proj
 `remove <PROJECT> ...`
 
-`remove **/*`
+`remove **/*.csproj`
```
I missed that one. It'll be in the next commit.

gforceg committed:
Update dotnet-sln.md

Remove specific globbing pattern from the remove command and explain that it 'Removes'.
Add globbing pattern place-holders to command descriptions
Fix globbing patterns in the examples

GuardRex commented on docs/core/tools/dotnet-sln.md:
```diff
@@ -65,18 +65,26 @@ Prints out a short help for the command.
-`dotnet sln todo.sln remove todo-app/todo-app.csproj`
+`dotnet sln todo.sln add todo-app/todo-app.csproj back-end/back-end.csproj`
+
+Remove multiple C# projects to a solution:
```
One more minor Nit ... use "from" ...

> Remove multiple C# projects from a solution:

gforceg reviewed (commented):
addresses the above suggestions / questions
also added a place holder to the add globbing place holder to the add command section.

gforceg committed:
fix typo

"remove to" should be "remove from"

gforceg commented:
@GuardRex Indeed they are. Thanks for helping me write this out correctly.
